tests/unit/test_abstract_fs.js

```diff
@@ -0,0 +1,137 @@
+    await expectCmd(afs.uploadWithExtraArgs,
+        ['/tmp/a', afs.resolve('file://host1/tmp', 'b'), '--exclude=*', '--include=*tfevents*'],
+        ['rsync -av /tmp/a host1:/tmp/b --exclude=* --include=*tfevents*']);
+
+    await expectCmd(afs.uploadWithExtraArgs,
+        ['/tmp/a', afs.resolve('/tmp', 'b'), '--exclude=*', '--include=*tfevents*'],
+        ['rsync -av /tmp/a /tmp/b --exclude=* --include=*tfevents*']);
+
+    await expectCmd(afs.uploadWithExtraArgs,
+        ['/tmp/a', afs.resolve('/tmp', 'a'), '--exclude=*', '--include=*tfevents*'],
+        []);
+
+    await expectCmd(afs.uploadWithExtraArgs,
+        ['/tmp', afs.resolve('s3://bucket/dir', 'a/'), '--exclude=*', '--include=*tfevents*'],
+        ['aws s3 sync /tmp s3://bucket/dir/a/ --exclude=* --include=*tfevents*']);
+
+    const jobid = 15;
+    await expectCmd(afs.uploadWithExtraArgs,
+        ['/home/workdir', afs.resolve('s3://bucket/dir', jobid.toString(), './tag:lang/'), '--exclude=*', '--include=*tfevents*'],
+        ['aws s3 sync /home/workdir s3://bucket/dir/15/tag:lang/ --exclude=* --include=*tfevents*']);
+}
+
+async function testSync() {
+    await expectCmd(afs.sync, [afs.resolve('/tmp/a'), afs.resolve('/tmp', 'b')],
+        ['rsync -av /tmp/a /tmp/b']);
+
+    await expectCmd(afs.sync, [afs.resolve('/tmp/a'), afs.resolve('/tmp', 'a')], 
+        ['rsync -av /tmp/a /tmp/a']);
+
+    await expectCmd(afs.sync, [afs.resolve('/tmp/a'), afs.resolve('file://host1/tmp', 'b')],
+        ['rsync -av /tmp/a host1:/tmp/b']);
+
+    await expectCmd(afs.sync, [afs.resolve('/tmp'), afs.resolve('s3://bucket/dir', 'a/')],
+        ['aws s3 sync /tmp s3://bucket/dir/a/']);
+
+    await expectCmd(afs.sync, [afs.resolve('s3://bucket/dir', 'a/'), afs.resolve('/workdir/a')],
+        ['aws s3 sync s3://bucket/dir/a/ /var/tmp/x',
+         'rsync -av /var/tmp/x /workdir/a',
+         'rm -r /var/tmp/x'
+         ]);
+
+    await expectCmd(afs.sync, [afs.resolve('s3://bucket/dir', 'a/'), afs.resolve('file://host2/workdir/a')],
+        ['aws s3 sync s3://bucket/dir/a/ /var/tmp/x',
+         'rsync -av /var/tmp/x host2:/workdir/a',
+         'rm -r /var/tmp/x'
+         ]);
+
+    await expectCmd(afs.sync, [afs.resolve('s3://bucket/dir', 'a/'), afs.resolve('file://host2/workdir/a'), '--extra=arg'],
+        ['aws s3 sync s3://bucket/dir/a/ /var/tmp/x --extra=arg',
+         'rsync -av /var/tmp/x host2:/workdir/a --extra=arg',
+         'rm -r /var/tmp/x'
+         ]);
+
+    const jobid = 15;
+    await expectCmd(afs.sync,
+        ['/home/workdir', afs.resolve('s3://bucket/dir', jobid.toString(), './tag:lang/'), '--exclude=* --include=*tfevents*'],
+        ['aws s3 sync /home/workdir s3://bucket/dir/15/tag:lang/ --exclude=* --include=*tfevents*']);
+}
+
+
+async function main() {
+    testResolve();
+    await testUpload();
+    await testUploadWithExtraArgs();
+    await testSync();
+}
+module.exports = main;
+if (!module.parent)
+    main();
```

Comment:
Thanks for adding tests! You should add this file to `tests/unit/index.js` so it will run automatically during `yarn test` (and during CI)

Reply:
Done.